packages/block-editor/src/components/inner-blocks/index.js

```diff
@@ -69,6 +83,31 @@ class InnerBlocks extends Component {
 				this.synchronizeBlocksWithTemplate();
 			}
 		}
+
+		// Sync with controlled blocks value from parent, if possible.
+		if ( this.isSyncingIncomingBlocks === innerBlocks ) {
```

Comment:
I think this logic is not going to work because innerBlocks will never have the same reference that isSyncingIncomingBlocks.
We call:
this.isSyncingIncomingBlocks = blocks;
replaceInnerBlocks( blocks );

When the replace operation is finished and we get new innerBlocks the reference of these inner blocks is not the same as the blocks we passed to replaceInnerBlocks. Because blocks are not saved as a tree in the state and the reference we get is new computation created by the getBlock selector.

Reply:
>When the replace operation is finished and we get new innerBlocks the reference of these inner blocks is not the same as the blocks we passed to replaceInnerBlocks. 

That's not ideal, could we get around it somehow? Maybe the reducer can keep a reference to the original array or something. In any case, this logic should not cause any issues if we leave it in until an enhancement is made to the store.